File: 0697-degree-of-an-array/0697-degree-of-an-array.java
class Solution {
    public int findShortestSubArray(int[] nums) {
        HashMap <Integer, Integer> map = new HashMap<>();
        HashMap <Integer, Integer> numIdxF = new HashMap<>();
        HashMap <Integer, Integer> numIdxL = new HashMap<>();
        for(int i=0; i<nums.length; i++) {
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);
            if(!numIdxF.containsKey(nums[i])) numIdxF.put(nums[i], i);
            numIdxL.put(nums[i], i);
        };
        int max = 0, res = Integer.MAX_VALUE;
        for(int key : map.keySet()){
            int temp = map.get(key);
            if(temp == max){
                int len = numIdxL.get(key) - numIdxF.get(key) + 1;
                if(res > len) res = len;
            } else if(temp > max){
                int len = numIdxL.get(key) - numIdxF.get(key) + 1;
                max = temp;
                res = len;
            }
        }
        return res;
    }
}

File: 0697-degree-of-an-array/0697-degree-of-an-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findShortestSubArray = function(nums) {
    const map = {};
    for(let i=0; i<nums.length; i++){
        if(!map[nums[i]]) {
            map[nums[i]] = {
                count: 0,
                left: i,
                right: i,
            };
        }
        map[nums[i]].count++;
        map[nums[i]].right = i;
    }
    let max = 0, res = nums.length;
    for(const key in map){
        let len = map[key].right - map[key].left + 1;
        if(map[key].count == max){
            res = Math.min(res, len);
        } else if( map[key].count > max){
            res = len;
            max = map[key].count;
        }
    }
    return res;
};